cache/nydus.go

```diff
@@ -0,0 +1,204 @@
+	layer.sr.mu.Lock()
+	defer layer.sr.mu.Unlock()
+
+	bootstrapDesc, err := layer.getBootstrap(ctx)
+	if err != nil {
+		return nil, errdefs.ErrNotFound
+	}
+
+	if compressionType == packer.CompressionTypeBootstrap {
+		return bootstrapDesc, err
+	} else if compressionType == packer.CompressionTypeBlob {
+		cachedBlob := layer.sr.GetString(keyCachedNydusBlob)
+		if cachedBlob == "" {
+			return nil, errdefs.ErrNotFound
+		}
+
+		if cachedBlob == emptySha256 {
+			// The nydus bootstrap layer exists but nydus blob layer does not,
+			// which means that the blob content of this layer is empty.
+			return nil, nil
+		}
+
+		if digest.Digest(cachedBlob).Validate() == nil {
+			desc, err := layer.sr.getCompressionBlob(ctx, compression.Parse(packer.CompressionTypeBlob))
+			if err != nil {
+				return nil, errdefs.ErrNotFound
+			}
+			return &desc, nil
+		}
+
+		return nil, errdefs.ErrNotFound
+	}
+
+	return nil, errdefs.ErrNotFound
+}
+
+// SetCache records nydus bootstrap/blob descriptor to cache,
+// desc == nil if nydus blob content is empty.
+func (layer *buildLayer) SetCache(ctx context.Context, compressionType packer.CompressionType, desc *ocispecs.Descriptor) error {
+	if compressionType == packer.CompressionTypeBlob {
+		blobDigest := emptySha256
+		if desc != nil {
+			blobDigest = desc.Digest.String()
+		}
+		if err := layer.sr.SetString(keyCachedNydusBlob, blobDigest, ""); err != nil {
+			return errors.Wrap(err, "cache nydus blob")
+		}
+	}
+	return nil
+}
+
+func (layer *buildLayer) ContentStore(ctx context.Context) content.Store {
+	return layer.sr.cm.ContentStore
+}
+
+func (layer *buildLayer) Mount(ctx context.Context) ([]mount.Mount, func() error, error) {
+	mountable, err := layer.sr.Mount(ctx, true, layer.sg)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return mountable.Mount()
+}
+
+func (layer *buildLayer) Backend(ctx context.Context) backend.Backend {
+	return nil
+}
+
+func isNydusBootstrap(desc ocispecs.Descriptor) bool {
+	if _, ok := desc.Annotations[nydusUtils.LayerAnnotationNydusBootstrap]; ok {
+		return true
+	}
+
+	return false
+}
+
+// Nydus blob layer can't be mixed with other compression types in the same image,
+// so if `source` is a nydus bootstrap layer, but the conversion target is other
+// compression type, we should do the forced conversion.
+func needsCompressionForNydus(source ocispecs.Descriptor, target compression.Type) bool {
+	if target == compression.NydusBlob {
+		return false
+	}
+
+	if source.Annotations == nil {
+		return false
+	}
+
+	return isNydusBootstrap(source)
+}
+
+func computeNydusBlobChain(ctx context.Context, sr *immutableRef, sg session.Group) ([]ocispecs.Descriptor, error) {
+	layers := []packer.Layer{}
+
+	sr.layerWalk(func(sr *immutableRef) {
+		layer := &buildLayer{
+			sr: sr,
+			sg: sg,
+		}
+		layers = append(layers, layer)
+	})
+
+	builder := os.Getenv("NYDUS_BUILDER")
+	if builder == "" {
+		builder = "nydus-image"
+	}
+	_packer, err := packer.New(packer.Option{
```

Comment:
What does the prefixed underscore mean?

Reply:
Just to avoid the variable name conflicts with the package name, already renamed to `p`.